src/distributions/uniform.rs

```diff
@@ -475,6 +475,139 @@ uniform_int_impl! { usize, isize, usize, isize, usize }
+macro_rules! uniform_simd_int_impl {
+    ($ty:ident, $unsigned:ident, $signed:ident, $u_scalar:ident) => {
+        // The "pick the largest zone that can fit in an `u32`" optimization
+        // is less useful here. Multiple lanes complicate things, we don't
+        // know the PRNG's minimal output size, and casting to a larger vector
+        // is generally a bad idea for SIMD performance. The user can still
+        // implement it manually.
+
+        // TODO: look into `Uniform::<u32x4>::new(0u32, 100)` functionality
+        //       perhaps `impl SampleUniform for $u_scalar`?
+        impl SampleUniform for $ty {
+            type Sampler = UniformInt<$ty>;
+        }
+
+        impl UniformSampler for UniformInt<$ty> {
+            type X = $ty;
+
+            #[inline] // if the range is constant, this helps LLVM to do the
+                      // calculations at compile-time.
+                      // TODO: test this ^ for SIMD
+            fn new<B1, B2>(low_b: B1, high_b: B2) -> Self
+                where B1: SampleBorrow<Self::X> + Sized,
+                      B2: SampleBorrow<Self::X> + Sized
+            {
+                let low = *low_b.borrow();
+                let high = *high_b.borrow();
+                assert!(low.lt(high).all(), "Uniform::new called with `low >= high`");
+                UniformSampler::new_inclusive(low, high - 1)
+            }
+
+            #[inline] // if the range is constant, this helps LLVM to do the
+                      // calculations at compile-time.
+            fn new_inclusive<B1, B2>(low_b: B1, high_b: B2) -> Self
+                where B1: SampleBorrow<Self::X> + Sized,
+                      B2: SampleBorrow<Self::X> + Sized
+            {
+                let low = *low_b.borrow();
+                let high = *high_b.borrow();
+                assert!(low.le(high).all(),
+                        "Uniform::new_inclusive called with `low > high`");
+                let unsigned_max = ::core::$u_scalar::MAX;
+
+                let range = $unsigned::from((high - low) + 1);
+                // The `range == 0` case is complicated with multiple lanes,
+                // avoid it.
+                assert!(range.eq($unsigned::splat(0)).none(),
+                        "Uniform::new_inclusive called with the full range");
+                let ints_to_reject = (unsigned_max - range + 1) % range;
+                let zone = unsigned_max - ints_to_reject;
+
+                UniformInt {
+                    low: low,
+                    // These are really $unsigned values, but store as $ty:
+                    range: $ty::from(range),
+                    zone: $ty::from(zone),
+                }
+            }
+
+            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X {
+                let range = $unsigned::from(self.range);
+                // TODO: ensure these casts are a no-op
+                let zone = $unsigned::from($signed::from(self.zone));
```

Comment:
I think simply doing 

```rust
$unsigned::from($signed::from(self.zone));
```

should work. We need multiple casts in the non-simd case because we're also changing the bit-width, but that's not the case here AFAICT.

Reply:
Err.. I mean

```rust
$unsigned::from(self.zone);
```